### PR TITLE
Join task fixes

### PR DESCRIPTION
**Pull request by VerstraeteBert**
A Join task will wrongfully be marked as COMPLETED_WITH_ERRORS if any task that is optional fails, regardless of the presence of a non-optional failure. 

This PR fixes that, adds a more comprehensive test suite for the join task, and refactors the join routine, additionally adds more null checks in case of unscheduled tasks.

Pull Request type
----
- [x] Bugfix

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/core/execution/tasks/Join.java [hunk 1]
```diff
@@ -47,41 +45,47 @@ public boolean execute(
...
                     joinOn.stream()
                             .map(name -> TaskUtils.appendIteration(name, task.getIteration()))
-                            .collect(Collectors.toList());
+                            .toList();
         }
+
+        boolean allTasksTerminal =
+                joinOn.stream()
+                        .map(workflow::getTaskByRefName)
+                        .allMatch(t -> t != null && t.getStatus().isTerminal());
```
Comment (mr-celo) [inline core/src/main/java/com/netflix/conductor/core/execution/tasks/Join.java:54]: Do we need to loop over twice? Can't we just piggy-back on the loop below?

**Review by v1r3n — APPROVED** [review]
